src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,258 @@
+//! This module specifies the type based interner for constants.
+//!
+//! After a const evaluation has computed a value, before we destroy the const evaluator's session
+//! memory, we need to extract all memory allocations to the global memory pool so they stay around.
+
+use rustc::ty::layout::LayoutOf;
+use rustc::ty;
+use rustc::mir::interpret::{
+    EvalResult, ErrorHandled,
+};
+use rustc::hir;
+use rustc::hir::def_id::DefId;
+use super::validity::RefTracking;
+use rustc_data_structures::fx::FxHashSet;
+
+use syntax::ast::Mutability;
+
+use super::{
+    ValueVisitor, MemoryKind, Pointer, AllocId, MPlaceTy, InterpError, Scalar,
+};
+use crate::const_eval::{CompileTimeInterpreter, CompileTimeEvalContext};
+
+struct InternVisitor<'rt, 'a: 'rt, 'mir: 'rt, 'tcx: 'a+'rt+'mir> {
+    /// previously encountered safe references
+    ref_tracking: &'rt mut RefTracking<(MPlaceTy<'tcx>, Mutability)>,
+    ecx: &'rt mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    mode: InternMode,
+    /// When encountering an immutable reference, we treat everything as immutable that is behind
+    /// it. If we encounter an `UnsafeCell` while `mutability == Immutable`, we ICE, because that
+    /// should have been prevented by const qualif
+    mutability: Mutability,
+    /// A list of all encountered relocations. After type-based interning, we traverse this list to
+    /// also intern allocations that are only referenced by a raw pointer or inside a union.
+    leftover_relocations: &'rt mut FxHashSet<AllocId>,
+}
+
+#[derive(Copy, Clone)]
+enum InternMode {
+    /// Mutable references don't change the `frozen` field to `false`
+    StaticMut,
+    /// Mutable references ICE
+    StaticOrConst,
+}
```

Comment:
Why is the `mutability` in the visitor not enough? `static mut` would start mutable while (frozen) `static` would start immutable and then the rules are the same, are they not?